Support running analyses on the public data

shankari opened the pull request:
Done so far:
- download data and store it to an temporary server (default `http://localhost:8080`). This allows us to delete and re-run analyses with multiple algorithms

USAGE:

Download the data (from this repo)

```
(emissioneval) e-mission-eval-public-data $ PYTHONPATH=. python bin/import_public_data.py http://cardshark.cs.berkeley.edu shankari@eecs.berkeley.edu car_scooter_brex_san_jose
```

Run the pipeline (from the server repo)

```
(emission) e-mission-server $ ./e-mission-py.bash bin/debug/intake_single_user.py -e ucb-sdb-ios-3
```

Next step:
- store the section ground truth as `mode_confirm` objects

Still need to figure out:
- how to store the trip ground truth? (`purpose_confirm`? `destination_confirm`?)
    - what about the trajectory?
- we are accessing the local server strictly through its public API, since we are outside the e-mission-server repository. This means that we can't:
   - load entries that have already passed through the pipeline once
   - delete entries (e.g. delete all the entries related to a particular spec once all the experiments are done)
   - run the pipeline automatically
   - reset the pipeline
   - ensure that the uuids on the local server are the same as the public datastore. Instead we re-register the phone labels and get new uuids and then change the `user_id` fields of the copied data.

None of these are showstoppers right now IMHO, so I am going to stick with this architecture. However, I just want to document the limitations. If any of them do become super important (e.g. loading entries that have passed through the pipeline), we can potentially address them by
1. adding new API methods, or 
1. splitting the monolithic server up into microservices and then including only the DB microservice code

shankari commented:
Working screenshots for this commit on two phones with different regimes for the mountain view -> San Jose library timeline

| iOS HAMFDC | android HAHFDC (begin) | android HAHFDC (end) |
| --------------| -----------------------| --------------------------- |
| ![Simulator Screen Shot - iPhone 7 - 2019-08-08 at 12 53 26](https://user-images.githubusercontent.com/2423263/62733522-a6dd2d00-b9db-11e9-806f-484257d3ed3c.png) | ![Simulator Screen Shot - iPhone 7 - 2019-08-08 at 12 54 04](https://user-images.githubusercontent.com/2423263/62733523-a775c380-b9db-11e9-8952-24e5551ce326.png) | ![Simulator Screen Shot - iPhone 7 - 2019-08-08 at 12 54 11](https://user-images.githubusercontent.com/2423263/62733524-a775c380-b9db-11e9-8e2f-5f457f32c4c8.png) |

shankari commented:
Working screenshots showing the ground truth showing up on the phone UI.
Note that in the case of the trip _to_ San Jose, the ground truth is displayed as "WALKING". This is because there are actually multiple ground truth sections - WALKING -> CAR -> WALKING. Since this ground truth is at the trip level, it expects only one ground truth per trip and if it finds multiple of them, it picks the last one. So in this case, it would pick the WALKING from the last section.

In the case of the BUS though, it does pick bus. Again probably because the walking ended too late to be a match or something. But at any rate, it does show that the ground truth is stored correctly.

| BUS (correct but unexpected)  |  WALKING (expected but incorrect) |
|----------------| ----------------------|
| ![Simulator Screen Shot - iPhone 7 - 2019-08-08 at 21 27 59](https://user-images.githubusercontent.com/2423263/62754448-b59c0200-ba24-11e9-8b3e-660351f125d3.png) | ![Simulator Screen Shot - iPhone 7 - 2019-08-08 at 21 28 13](https://user-images.githubusercontent.com/2423263/62754449-b59c0200-ba24-11e9-8c78-56bc79f1953a.png) |

shankari commented:
Just to verify that the mode confirm is in fact stored correctly, let's verify
this from the timestamps. Again, we don't really expect this to work, since the
tested use case is to set the `mode_confirm` unimodally from the UI instead of
multi-modally from out-of-band ground truth, but I want to check that this
expectation of incorrectness is not masking anything more serious.

And everything looks good. Details below...

For the trip to San Jose

```
potentialCandidates are
1565026080.3062959 -> 1565026125.9911141 WALKING logged at 1565026125.9911141
1565026126.007808 -> 1565027744.871016 CAR logged at 1565027744.871016
1565027744.885445 -> 1565027779.040153 WALKING logged at 1565027779.040153

Returning mostRecentEntry 1565027744.885445 -> 1565027779.040153 WALKING logged at 1565027779.040153
```

For the trip from San Jose

```
[Log] potentialCandidates are
1565043965.8569732 -> 1565044035.215605 WALKING logged at 1565044035.215605
1565044035.23178 -> 1565044463.004527 STOPPED logged at 1565044463.004527
1565044463.0175738 -> 1565046534.8845382 BUS logged at 1565046534.8845382
[Log] Returning mostRecentEntry 1565044463.0175738 -> 1565046534.8845382 BUS logged at 1565046534.8845382
```

Checking even further, we find all the entries in our list, in particular

```
30 "1565042553.9210582 -> 1565042612.198419 WALKING logged at 1565042612.198419"
31 "1565042612.2229319 -> 1565043965.839017 ESCOOTER logged at 1565043965.839017"
32 "1565043965.8569732 -> 1565044035.215605 WALKING logged at 1565044035.215605"
33 "1565044035.23178 -> 1565044463.004527 STOPPED logged at 1565044463.004527"
34 "1565044463.0175738 -> 1565046534.8845382 BUS logged at 1565046534.8845382"
35 "1565046534.919526 -> 1565047154.3463578 WALKING logged at 1565047154.3463578"
```

This particular trip has start and end

[1565042674.0040407, "2019-08-05T15:04:34.004041-07:00",
 1565047109.994664, "2019-08-05T16:18:29.994664-07:00"]

The first WALKING and ESCOOTER started before the detected trip start (due to
the known issue with iOS delayed start tracking). The last WALKING ends after
the trip end was detected, only by a minute or so, so is not displayed.

jf87 commented on bin/import_public_data.py:
```diff
@@ -26,8 +26,29 @@ def save_to_db(db_url, phone_view):
+                    for sr in tr["evaluation_section_ranges"]:
+                        print(16 * ' ', 30 * "~")
+                        print(16 * ' ',sr["trip_id"], sr.keys())
+                        gt_leg = phone_view.spec_details.get_ground_truth_for_leg(sr["trip_id_base"])
```
You should add a check here:
`if gt_leg != None:`
Otherwise it crashes if there is no ground truth available.
This just happened for me when importing unimodal_trip_car_bike_mtv_la

shankari replied:
hm, but there should always be ground truth available since we generate the timeline data from the spec. Let me try re-running...

shankari replied:
Ah this is because there are two separate walking sections for each trip - one to walk to the vehicle and one to walk from the vehicle. In the long term, I should ensure that the second walk section is `walk_end` instead of `walk_start`. But now we have collected data with this label and I don't want to relabel, so I will use a minimally more sophisticated check.

```
walk_start 2 ['walk_start', 'walk_start']
walk_start 2 ['walk_start', 'walk_start']
```

shankari replied:
I am not sure why this is not affecting the `car_scooter_...` timeline, since there are multiple `walk_start`s in it too, including two in the first `freeway` trip.

```
$ grep -r walk_start spec_creation/*.filled.json
spec_creation/car_scooter_brex_san_jose.filled.json:          "id": "walk_start",
spec_creation/car_scooter_brex_san_jose.filled.json:          "id": "walk_start",
spec_creation/car_scooter_brex_san_jose.filled.json:          "id": "walk_start",
spec_creation/train_bus_ebike_mtv_ucb.filled.json:          "id": "walk_start",
spec_creation/unimodal_roundtrip.filled.json:          "id": "walk_start",
spec_creation/unimodal_roundtrip.filled.json:          "id": "walk_start",
spec_creation/unimodal_roundtrip.filled.json:          "id": "walk_start",
spec_creation/unimodal_roundtrip.filled.json:          "id": "walk_start",
```

shankari replied:
Ah, it's because we iterate through the trips and the second trip has only one `walk_start`

```
walk_start 2 ['walk_start', 'walk_start']
walk_start 1 ['walk_start']
```

shankari replied:
ok so fix this I can take one of two approaches:
- ensure that all legs in a trip, including the automatically added shim legs are unique going forward. Fix the labels on the existing ground truth segments to match.
    - Note that this is a bit tricky because we may have multiple transfers between similar modes; for e.g. if we have a walk -> bus -> train -> bus -> train -> walk timeline, we will have two bus -> train transfers. We would then need to check for that and make the tripid unique while autofilling the spec.
- in addition to the leg_id, we can pass in the trip_id and the locations for the leg, and return  the leg with the greatest overlap between the spatial ground truth and the actual points.

shankari replied:
I think that overall, it would be easiest for to ensure that the shim leg ids are unique going forward. This is because I don't want to involve any analysis in figuring out the "ground truth"; that should be as immutable as possible.

shankari commented:
After regenerating all the timelines, I will fix the labels on the ground truth.
The main changes are:
- `unimodal_trip_car_bike_mtv_la`
   - change the second `walk_start` to `walk_end` (DONE)
      ```
      PYTHONPATH=. python bin/import_public_data.py http://cardshark.cs.berkeley.edu shankari@eecs.berkeley.edu unimodal_trip_car_bike_mtv_la
      ...
      walk_end_0 dict_keys(['trip_id', 'trip_id_base', 'trip_run', 'start_ts', 'end_ts', 'duration', 'battery_df', 'location_df', 'motion_activity_df', 'transition_df'])
      About to post messages {'user': 'ucb-sdb-ios-4', 'the_entry': {'metadata': {'write_ts': 1564358080.2698169, 'read_ts': 1564358080.2698169, 'time_zone': 'America/Los_Angeles', 'type': 'message', 'key': 'manual/mode_confirm', 'platform': 'ios'}, 'data': {'label': 'WALKING', 'start_ts': 1564358050.1955462, 'end_ts': 1564358080.2698169}}}
      response = <Response [200]>
      ```

- `car_scooter_brex_san_jose`
   - convert the second leg of the first trip to `walk_end` (DONE)
   - convert `tt_ESCOOTER_BUS` to `tt_city_escooter_city_bus_rapid_transit` (DONE)
   - convert `wait_for_BUS` to `wait_for_city_bus_rapid_transit` (DONE)
  
    ```
    PYTHONPATH=. python bin/import_public_data.py http://cardshark.cs.berkeley.edu shankari@eecs.berkeley.edu car_scooter_brex_san_jose.filled.json
                ....
                     tt_city_escooter_city_bus_rapid_transit_0 dict_keys(['trip_id', 'trip_id_base', 'trip_run', 'start_ts', 'end_ts', 'duration', 'battery_df', 'location_df', 'motion_activity_df', 'transition_df'])
                  About to post messages {'user': 'ucb-sdb-ios-4', 'the_entry': {'metadata': {'write_ts': 1565211770.6105618, 'read_ts': 1565211770.6105618, 'time_zone': 'America/Los_Angeles', 'type': 'message', 'key': 'manual/mode_confirm', 'platform': 'ios'}, 'data': {'label': 'WALKING', 'start_ts': 1565211705.4192839, 'end_ts': 1565211770.6105618}}}
                  response = <Response [200]>
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                 wait_for_city_bus_rapid_transit_0 dict_keys(['trip_id', 'trip_id_base', 'trip_run', 'start_ts', 'end_ts', 'duration', 'battery_df', 'location_df', 'motion_activity_df', 'transition_df'])
                 About to post messages {'user': 'ucb-sdb-ios-4', 'the_entry': {'metadata': {'write_ts': 1565211919.974773, 'read_ts': 1565211919.974773, 'time_zone': 'America/Los_Angeles', 'type': 'message', 'key': 'manual/mode_confirm', 'platform': 'ios'}, 'data': {'label': 'STOPPED', 'start_ts': 1565211770.626648, 'end_ts': 1565211919.974773}}}
                 ...
                 response = <Response [200]>
    ```

shankari commented:
And finally, `train_bus_ebike_mtv_ucb`
- convert `wait_for_TRAIN` to `wait_for_commuter_rail_aboveground` (DONE)
- convert `tt_TRAIN_SUBWAY` to `tt_commuter_rail_aboveground_subway_underground` (DONE)
- convert `wait_for_SUBWAY` to `wait_for_subway_underground` (DONE)
- convert `wait_for_BUS` to `wait_for_city_bus_short` (DONE)
- convert `walk_start` to `walk_end` (DONE)
- convert `tt_BICYCLING_BUS` to `tt_ebike_bikeshare_urban_long_express_bus` (DONE)
- convert `wait_for_BUS` to `wait_for_express_bus` (DONE)
- convert `tt_LIGHT_RAIL_TRAIN` to `tt_light_rail_below_above_ground_commuter_rail_with_tunnels` (DONE)
- convert `wait_for_TRAIN` to `wait_for_commuter_rail_with_tunnels` (DONE)

shankari commented:
@jf87 fixed a much more complicated, but I think, much more principled fashion.
LMK if you find any other issues.

jf87 commented:
Thanks a lot 👍 

Can I just re-run the import or better clean the users from the DB?

shankari commented:
probably better to clean the users and re-run

jf87 commented:
Sorry to bother you again.

Running
`PYTHONPATH=. python bin/import_public_data.py http://cardshark.cs.berkeley.edu shankari@eecs.berkeley.edu unimodal_trip_car_bike_mtv_la
`
results now in:

`{'transition': 'STOP_EVALUATION_PERIOD', 'trip_id': 'HAHFDC v/s MAHFDC:MAHFDC', 'spec_id': 'unimodal_trip_car_bike_mtv_la', 'device_manufacturer': 'motorola', 'device_model': 'Nexus 6', 'device_version': '6.0.1', 'ts': 1565578932.067, 'write_ts': 1565578932.123} {'transition': 'STOP_EVALUATION_PERIOD', 'trip_id': 'HAHFDC v/s MAHFDC:MAHFDC', 'spec_id': 'unimodal_trip_car_bike_mtv_la', 'device_manufacturer': 'motorola', 'device_model': 'Nexus 6', 'device_version': '6.0.1', 'ts': 1565578932.067, 'write_ts': 1565578932.123}
{'HAHFDC v/s HAMFDC:HAMFDC', 'HAHFDC v/s MAHFDC:MAHFDC'}
Traceback (most recent call last):
  File "bin/import_public_data.py", line 68, in <module>
    for sid in args.spec_list]
  File "bin/import_public_data.py", line 68, in <listcomp>
    for sid in args.spec_list]
  File "bin/import_public_data.py", line 9, in load_from_spec_id
    pv = eipv.PhoneView(sd)
  File "/Users/nle5268/Sync/src/e-mission-eval-public-data/emeval/input/phone_view.py", line 46, in __init__
    self.fill_evaluation_ranges()
  File "/Users/nle5268/Sync/src/e-mission-eval-public-data/emeval/input/phone_view.py", line 264, in fill_evaluation_ranges
    self.spec_details.eval_end_ts)
  File "/Users/nle5268/Sync/src/e-mission-eval-public-data/emeval/input/phone_view.py", line 169, in transitions_to_ranges
    assert s["transition"] == start_tt or s["transition"] == start_ti, "Start transition has %s transition" % s["transition"]
AssertionError: Start transition has STOP_EVALUATION_PERIOD transition
`

Any ideas?

jf87 commented:
And for multimodal_trip_mtv_ucb:

`About to retrieve messages using {'user': 'shankari@eecs.berkeley.edu', 'key_list': ['config/evaluation_spec'], 'start_time': 0, 'end_time': 1565598457}
response = <Response [200]>
Found 11 entries
After iterating over 11 entries, entry not found
Traceback (most recent call last):
  File "bin/import_public_data.py", line 68, in <module>
    for sid in args.spec_list]
  File "bin/import_public_data.py", line 68, in <listcomp>
    for sid in args.spec_list]
  File "bin/import_public_data.py", line 8, in load_from_spec_id
    sd = eisd.SpecDetails(datastore_url, email, spec_id)
  File "/Users/nle5268/Sync/src/e-mission-eval-public-data/emeval/input/spec_details.py", line 14, in __init__
    self.populate_spec_details(self.curr_spec_entry)
  File "/Users/nle5268/Sync/src/e-mission-eval-public-data/emeval/input/spec_details.py", line 57, in populate_spec_details
    self.curr_spec_wrapper = self.curr_spec_entry["data"]
TypeError: 'NoneType' object is not subscriptable
`

shankari commented:
> Running
PYTHONPATH=. python bin/import_public_data.py http://cardshark.cs.berkeley.edu shankari@eecs.berkeley.edu unimodal_trip_car_bike_mtv_la
results now in:

We collected some more data for this yesterday and it looks like maybe not everything got uploaded from the phones? I will take a look

shankari commented:
> And for multimodal_trip_mtv_ucb:

I am not sure you are using the correct spec id here. The UCB timeline is `train_bus_ebike_mtv_ucb` 

```
(emissioneval) bash-3.2$ PYTHONPATH=. python bin/import_public_data.py http://cardshark.cs.berkeley.edu shankari@eecs.berkeley.edu train_bus_ebike_mtv_ucb
After iterating over 11 entries, entry found
Found spec = Multimodal multi-train, multi-bus, ebike trip to UC Berkeley
Evaluation ran from 2019-07-16T00:00:00-07:00 -> 2019-09-30T00:00:00-07:00
-------------------- About to read transitions from server --------------------
Reading data for android phones
Loading transitions for phone ucb-sdb-android-1
About to retrieve messages using {'user': 'ucb-sdb-android-1', 'key_list': ['manual/evaluation_transition'], 'start_time': 1563260400, 'end_time': 1569826800}
response = <Response [200]>
Found 70 entries
```

jf87 commented:
Oh, sorry. I am running train_bus_ebike_mtv_ucb now and it seems to work.

shankari commented:
> We  collected some more data for this yesterday and it looks like maybe not  everything got uploaded from the phones? I will take a look

@jf87 Ok, so this should be fixed now. What happened was that we collected some more data yesterday (note that there is now also a `'HAHFDC v/s MAHFDC` for android and `'HAHFDC v/s HAMFDC' for iOS). However, the first attempt at data collection failed because the person collecting the data forgot to hit "Next" when he got out of the car at the Los Altos library. So we decided to abandon that attempt and start over.

So I deleted the data for that run, but the delete script was based on `data.ts`, not `metadata.write_ts`, so it missed the "end of experiment" transition for that run. I just deleted it and it should work now.

jf87 commented:
Cool. I just imported now also unimodal_trip_car_bike_mtv_la.
Can you post again the link to the specs?
I think there were one or two more.

shankari commented:
@jf87 there are only 3 timelines :(
https://github.com/e-mission/e-mission-eval-public-data/tree/master/spec_creation/final_sfbayarea

Note that for `car_scooter_brex_san_jose` and `unimodal_trip_car_bike_mtv_la`, we have the full set of comparisons - we have high accuracy high frequency comparisons to lowering frequency (high accuracy medium frequency) and lowering accuracy (medium accuracy high frequency). Would be interesting to see which regime does better with your algorithm...
